src/components/remote_control/src/resource_allocation_manager_impl.cc

```diff
@@ -0,0 +1,215 @@
+                          << module_type);
+    return AcquireResult::ALLOWED;
+  }
+
+  if (IsModuleTypeRejected(module_type, app_id)) {
+    LOG4CXX_DEBUG(logger_,
+                  "Driver disallowed app: " << app_id << " to acquire "
+                                            << module_type);
+    return AcquireResult::REJECTED;
+  }
+
+  const mobile_apis::HMILevel::eType acquiring_app_hmi_level =
+      acquiring_app->hmi_level();
+
+  if (mobile_apis::HMILevel::HMI_FULL != acquiring_app_hmi_level) {
+    LOG4CXX_DEBUG(
+        logger_,
+        "Aquiring resources is not allowed in HMI level: "
+            << application_manager::MessageHelper::StringifiedHMILevel(
+                   acquiring_app_hmi_level) << ". App: " << app_id
+            << " is disallowed to acquire " << module_type);
+    return AcquireResult::REJECTED;
+  }
+
+  switch (current_access_mode_) {
+    case hmi_apis::Common_RCAccessMode::AUTO_DENY: {
+      LOG4CXX_DEBUG(logger_,
+                    "Current access_mode is AUTO_DENY. "
+                        << "App: " << app_id << " is disallowed to acquire "
+                        << module_type);
+      return AcquireResult::IN_USE;
+    }
+    case hmi_apis::Common_RCAccessMode::ASK_DRIVER: {
+      LOG4CXX_DEBUG(logger_,
+                    "Current access_mode is ASK_DRIVER. "
+                    "Driver confirmation is required for app: "
+                        << app_id << " to acquire " << module_type);
+      return AcquireResult::ASK_DRIVER;
+    }
+    case hmi_apis::Common_RCAccessMode::AUTO_ALLOW: {
+      LOG4CXX_DEBUG(logger_,
+                    "Current access_mode is AUTO_ALLOW. "
+                        << "App: " << app_id << " is allowed to acquire "
+                        << module_type);
+
+      allocated_resources_[module_type] = app_id;
+      return AcquireResult::ALLOWED;
+    }
+    default: { DCHECK_OR_RETURN(false, AcquireResult::IN_USE); }
+  }
+}
+
+void ResourceAllocationManagerImpl::SetResourceState(
+    const std::string& module_type,
+    const uint32_t app_id,
+    const ResourceState::eType state) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  LOG4CXX_DEBUG(logger_,
+                "Setting state for " << module_type << " by app_id " << app_id
+                                     << " to state " << state);
+  const AllocatedResources::const_iterator allocated_it =
+      allocated_resources_.find(module_type);
+
+  const std::string status = allocated_resources_.end() != allocated_it
+                                 ? " acquired "
+                                 : " not acquired ";
+  UNUSED(status);
+  LOG4CXX_DEBUG(logger_,
+                "Resource " << module_type << " is " << status
+                            << " Owner application id is "
+                            << allocated_it->second
+                            << " Changing application id is " << app_id);
+
+  resources_state_[module_type] = state;
```

Comment:
Because SetResourceState can be called by multiple commands, do there need to be locks around setting and reading the resources_state_ map?

Reply:
Same with, allocated_resources_. Does this need locks around usage?

Reply:
@JackLivio race condition will be checked and additional locks will be seted